### ISSUE TITLE
Improve TH support in GetModIfaceFromDisk

### ISSUE THREAD

**Issue by pepeiborra** [haskell/ghcide#660]
* Reduce interface reloading by implementing stability checks
* Add test cases of interface reloading in the presence of TH 

_See https://github.com/digital-asset/ghcide/pull/645/files/49b0d9ac65399edf82a7a9cbbb8d8b5420458d8d_

**Comment by pepeiborra** [comment]
Coming back to this ticket, a stability check is simply being able to tell whether a module source file and all its dependencies have changed since the module interface file was computed. If none has changed then we have the "stability" property, otherwise we don't. If this module or any of the dependencies use TH, stability is a required condition to reuse interface files.

In ghcide interface loading happens in the `GetModIfaceFromDisk` rule.
When an evaluation of the rule kicks in, it's either the first time that the rule has been demanded, or a re-run due to a change in dependencies.

1. If it's a re-run, there are two cases:
  1a. The ModSummary has changed. In this case the module source file must have changed and therefore stability does not apply.
  1b. A dependency has changed. There are some corner cases, like when a dependency changes in a way that the associated interface file doesn't, but I think it's fair to assume that stability doesn't hold.

2. If it is the first time the rule is computed, like at startup, then it's very possible that stability holds. Moreover, failing to detect stability means we will need to regenerate all the interface files for modules that use TH directly or depend on a module that does.

To summarise: case 1 is fine, but case 2 needs some attention to avoid doing too much work when TH is used.

**Comment by pepeiborra** [comment]
This is fixed
